<a href="https://colab.research.google.com/github/lacey-conrad/WGCNA-work/blob/main/Gene_coexp_NN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

import tensorflow


#from sklearn.model_selection import train_test_split
#from sklearn.preprocessing import StandardScaler
#from sklearn import metrics
#from sklearn.metrics import confusion_matrix, accuracy_score
#from sklearn.preprocessing import LabelEncoder

#from tensorflow.keras import Sequential
#from tensorflow.keras.layers import Dense
#from tensorflow import keras
#from tensorflow.keras.layers import LeakyReLU

# 1. Reading in the dataset:

In [3]:
df = pd.read_table('thermo_capacity_counts.txt')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [6]:
df_gastroc = pd.read_csv('gastroc_raw_counts.csv')

In [ ]:
#from google.colab import drive
#drive.mount('/content/drive')

In [ ]:
#/content/thermo_capacity_counts.txt

In [ ]:
#df_copy = df.copy()

In [11]:
df_gastroc.head()

Unnamed: 0 LN.F1.113.2be LN.F1.113.aa LN.F1.113.ae LN.F1.113.ag  \
0  LOC121821266             0            0            0            0   
1  LOC102911281             0            0            0            0   
2  LOC102926907             0            0            0            0   
3  LOC121821617             0            0            0            1   
4  LOC102925994             3            4            3            4   

  LN.F1.113.ak LN.F1.113.am LN.F1.113.b LN.F1.113.h LN.F1.113.p  ...  \
0            0            0           0           0           0  ...   
1            0            0           0           0           0  ...   
2            0            0           0           0           0  ...   
3            0            0           0           0           0  ...   
4            7            7           9           2           6  ...   

  ME.F1.145.ad ME.F1.145.d ME.F1.145.g ME.F1.145.i ME.F1.145.w ME.F1.146.aa  \
0            0           0           0           0           0            0   
1            0           0           0           0           0            0   
2            0           0           0           0           0            0   
3            0           1           0           1           0            0   
4            3           9           5           6          11            2   

  ME.F1.146.ab ME.F1.146.ac ME.F1.146.be ME.F1.146.bf  
0            0            0            0            0  
1            0            0            0            0  
2            0            0            0            0  
3            0            1            1            0  
4            2            4            2            5  

[5 rows x 48 columns]

In [9]:
df_gastroc = df_gastroc.T

In [16]:
df_gastroc.columns = df_gastroc.iloc[0]

In [18]:
df_gastroc = df_gastroc.iloc[1: , :]

In [19]:
df_gastroc.head()

Unnamed: 0    LOC121821266 LOC102911281 LOC102926907 LOC121821617  \
LN.F1.113.2be            0            0            0            0   
LN.F1.113.aa             0            0            0            0   
LN.F1.113.ae             0            0            0            0   
LN.F1.113.ag             0            0            0            1   
LN.F1.113.ak             0            0            0            0   

Unnamed: 0    LOC102925994 LOC102925679 LOC121821541 LOC102925362  \
LN.F1.113.2be            3            0            3            3   
LN.F1.113.aa             4            0            1            0   
LN.F1.113.ae             3            0            1            2   
LN.F1.113.ag             4            0            0            0   
LN.F1.113.ak             7            0            0            3   

Unnamed: 0    LOC121821593 LOC102924750  ... EI127_mgp04 EI127_mgt17  \
LN.F1.113.2be            0            0  ...       11886         276   
LN.F1.113.aa             0            0  ...       12638         371   
LN.F1.113.ae             0            0  ...        9748         272   
LN.F1.113.ag             0            0  ...       16001         495   
LN.F1.113.ak             1            0  ...       13241         337   

Unnamed: 0    EI127_mgt18 EI127_mgt19 EI127_mgp03 EI127_mgp02 EI127_mgt20  \
LN.F1.113.2be         122          92        6169        5420        2144   
LN.F1.113.aa          207          81        7024        6495        1673   
LN.F1.113.ae          135          67        5841        5400        1414   
LN.F1.113.ag          230         116        9873        7522        2011   
LN.F1.113.ak          180          91        7773        6657        1650   

Unnamed: 0    EI127_mgp01 EI127_mgt21 EI127_mgt22  
LN.F1.113.2be       16304         173        1511  
LN.F1.113.aa        16561         142        1280  
LN.F1.113.ae        13018         103         955  
LN.F1.113.ag        21779         196        1599  
LN.F1.113.ak        16713         109        1130  

[5 rows x 32829 columns]

In [8]:
#pd.isnull(df_gastroc).count()

Unnamed: 0        32829
LN.F1.113.2be     32829
LN.F1.113.aa      32829
LN.F1.113.ae      32829
LN.F1.113.ag      32829
LN.F1.113.ak      32829
LN.F1.113.am      32829
LN.F1.113.b       32829
LN.F1.113.h       32829
LN.F1.113.p       32829
LN.F1.132.af      32829
LN.F1.132.aj      32829
LN.F1.132.aq.2    32829
LN.F1.132.ar      32829
LN.F1.132.as      32829
LN.F1.132.at      32829
LN.F1.132.aw      32829
LN.F1.132.ax      32829
LN.F1.132.ay      32829
LN.F1.132.ba      32829
LN.F1.132.bc      32829
LN.F1.132.be.2    32829
LN.F1.132.bh      32829
LN.F1.132.t       32829
ME.F1.134.a       32829
ME.F1.134.b       32829
ME.F1.136.h       32829
ME.F1.136.i       32829
ME.F1.136.j       32829
ME.F1.139.af      32829
ME.F1.139.ag      32829
ME.F1.141.at      32829
ME.F1.141.q       32829
ME.F1.141.r       32829
ME.F1.141.t       32829
ME.F1.141.u       32829
ME.F1.141.w       32829
ME.F1.144.u       32829
ME.F1.145.ad      32829
ME.F1.145.d       32829
ME.F1.145.g       32829
ME.F1.145.i     

In [ ]:
#df.describe()

In [21]:
df_gastroc = df_gastroc.reset_index()

In [22]:
df_gastroc.head()

Unnamed: 0          index LOC121821266 LOC102911281 LOC102926907 LOC121821617  \
0           LN.F1.113.2be            0            0            0            0   
1            LN.F1.113.aa            0            0            0            0   
2            LN.F1.113.ae            0            0            0            0   
3            LN.F1.113.ag            0            0            0            1   
4            LN.F1.113.ak            0            0            0            0   

Unnamed: 0 LOC102925994 LOC102925679 LOC121821541 LOC102925362 LOC121821593  \
0                     3            0            3            3            0   
1                     4            0            1            0            0   
2                     3            0            1            2            0   
3                     4            0            0            0            0   
4                     7            0            0            3            1   

Unnamed: 0  ... EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03 EI127_mgp02  \
0           ...         276         122          92        6169        5420   
1           ...         371         207          81        7024        6495   
2           ...         272         135          67        5841        5400   
3           ...         495         230         116        9873        7522   
4           ...         337         180          91        7773        6657   

Unnamed: 0 EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22 Index  
0                 2144       16304         173        1511     1  
1                 1673       16561         142        1280     2  
2                 1414       13018         103         955     3  
3                 2011       21779         196        1599     4  
4                 1650       16713         109        1130     5  

[5 rows x 32831 columns]

In [24]:
df_gastroc = df_gastroc.rename(columns={'index': 'Treatment'})

In [27]:
df_gastroc.head()

Unnamed: 0      Treatment LOC121821266 LOC102911281 LOC102926907 LOC121821617  \
0           LN.F1.113.2be            0            0            0            0   
1            LN.F1.113.aa            0            0            0            0   
2            LN.F1.113.ae            0            0            0            0   
3            LN.F1.113.ag            0            0            0            1   
4            LN.F1.113.ak            0            0            0            0   

Unnamed: 0 LOC102925994 LOC102925679 LOC121821541 LOC102925362 LOC121821593  \
0                     3            0            3            3            0   
1                     4            0            1            0            0   
2                     3            0            1            2            0   
3                     4            0            0            0            0   
4                     7            0            0            3            1   

Unnamed: 0  ... EI127_mgt17 EI127_mgt18 EI127_mgt19 EI127_mgp03 EI127_mgp02  \
0           ...         276         122          92        6169        5420   
1           ...         371         207          81        7024        6495   
2           ...         272         135          67        5841        5400   
3           ...         495         230         116        9873        7522   
4           ...         337         180          91        7773        6657   

Unnamed: 0 EI127_mgt20 EI127_mgp01 EI127_mgt21 EI127_mgt22 Index  
0                 2144       16304         173        1511     1  
1                 1673       16561         142        1280     2  
2                 1414       13018         103         955     3  
3                 2011       21779         196        1599     4  
4                 1650       16713         109        1130     5  

[5 rows x 32831 columns]

In [ ]:

X = df.drop(labels=['CustomerId', 'Surname', 'RowNumber', 'Exited'], axis = 1)
y = df.Exited

In [ ]:
X.info()

In [ ]:
# Using one hot encoding on geography and gender to convert categorial variables to numeric variables.
X = pd.get_dummies(X, columns=['Geography', 'Gender'])

In [ ]:
# Checking to see how the data is spread amongst the feature subcategories:
# **the floats data type was not counted since each value has a count of 1 (generally)
for i in list(X.columns[X.dtypes=='int64']):  
    print(i)
    print(X[i].value_counts().sort_index().head())
    print()

In [ ]:
for i in list(X.columns[X.dtypes=='uint8']):  
    print(i)
    print(X[i].value_counts().sort_index().head())
    print()

In [ ]:
sns.pairplot(data = df_copy)

In [ ]:
plt.subplots(figsize=(20,15))
sns.heatmap(df_copy.corr(),annot=True,square=True)

# Insights from Pairplot and Correlation Matrix:

- No noticable correlations were observed in the pairplot.

- Age seems to be positively related to Exited status.

- The number of products a person has seems to be negatively related to their balance.

- Most other relationships were very small.

In [ ]:
fig = sns.FacetGrid(df,hue='Exited',aspect=4)
fig.map(sns.kdeplot,'Age',shade=True)
fig.add_legend()

In [ ]:
fig = sns.FacetGrid(df,hue='Exited',aspect=4)
fig.map(sns.kdeplot,'CreditScore',shade=True)
fig.add_legend()

In [ ]:
fig = sns.FacetGrid(df,hue='Exited',aspect=4)
fig.map(sns.kdeplot,'Tenure',shade=True)
fig.add_legend()

In [ ]:
sns.set(rc={'figure.figsize':(20,6)})
sns.countplot(x = df_copy['NumOfProducts'], data=df_copy, hue=df_copy['Exited'])
plt.title('Count Plot of "Number of Products" for both exit conditions')

In [ ]:
sns.set(rc={'figure.figsize':(20,6)})
sns.countplot(x = df_copy['HasCrCard'], data=df_copy, hue=df_copy['Exited'])
plt.title('Count Plot of "Has Credit Card" for both exit conditions')

In [ ]:
sns.set(rc={'figure.figsize':(20,6)})
sns.countplot(x = df_copy['IsActiveMember'], data=df_copy, hue=df_copy['Exited'])
plt.title('Count Plot of "Is an Active Member" for both exit conditions')

In [ ]:
sns.set(rc={'figure.figsize':(20,6)})
sns.countplot(x = df_copy['Geography'], data=df_copy, hue=df_copy['Exited'])
plt.title('Count Plot of exit condition per country')

In [ ]:
sns.set(rc={'figure.figsize':(20,6)})
sns.countplot(x = df_copy['Gender'], data=df_copy, hue=df_copy['Exited'])
plt.title('Count Plot of exit condition for females and males')

# Insights from bivariate analysis:

- There was now significant difference between the credit scores of those who exited the bank and those who did not.

- There is not significant difference between the tenure of those who exited the bank and those who did not.

- The age of those who did not exit the bank was older than the age of those who did exit the bank.  In particular, those in their late 40s seemed very likely not to have exited the bank.

- The likelihood someone would leave the bank decreased with increasing number of products the person has.  Having 1 product with the bank contained the highest likelihood someone would leave the bank, followed by 2, 3, and then 4.

- The likelihood someone would leave the bank if they had a credit card through the bank was roughly equal to those who did not have a credit card with the bank.

- Those customers from Germany were the most likely out of the three countries surveyed to leave the bank.  

- Women were more likely to leave the bank than men.

In [ ]:
_, ax = plt.subplots(1, 3, figsize=(18, 6))
plt.subplots_adjust(wspace=0.3)
sns.swarmplot(x = "NumOfProducts", y = "Age", hue="Exited", data = df_copy, ax= ax[0])
sns.swarmplot(x = "HasCrCard", y = "Age", data = df_copy, hue="Exited", ax = ax[1])
sns.swarmplot(x = "IsActiveMember", y = "Age", hue="Exited", data = df_copy, ax = ax[2])

# 4. Distinguishing the feature and target set and dividing the data into training and test sets:

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state = 42, stratify = y)

In [ ]:
X.head()

# 5. Normalizing the training and testing data:

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 6. Initialize and build the model.  Identify the points of improvement and implement the same, 

# 7. Predict the results using a 0.5 threshold, and

# 8. Print the accuracy score and confusion matrix.

** I assumed that by "points of improvement" the intent is to test different parameters and hyperparameters to see what produces the most accurate model.  Therefore, I created 3 different models with varying sets of parameters to see how they effect the accuracy score of the model.  I will keep the output layer set to 1 node, with a sigmoid activation function, since this will give the predictions as well as probabilities.  A 1 node output layer is utilized since the classification is binary.

In [ ]:
# Model#1
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 10

In [ ]:
# Initializing the ANN
model = Sequential()

# Adding the input layers and hidden layers:
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))

# Adding the output layer:
model.add(Dense(1, activation = 'sigmoid'))

# Compiling the ANN:
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])

# Training the ANN:
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 10, verbose=1) 

In [ ]:
# Making predictions:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
# Predictions of the response variable (whether or not the customer will exit the bank)
y_pred

In [ ]:
# Evaluating the model:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#2 - more nodes
# Input layer: 50 nodes
# Hidden layer 1: 500 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 10

In [ ]:
model = Sequential()
model.add(Dense(50, activation='relu'))
model.add(Dense(500, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 10, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#3 - larger batch size
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 10
# Batch size: 500

In [ ]:
model = Sequential()
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 10, batch_size = 500, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

In [ ]:
# Model#4 - More epochs
# Input layer: 32 nodes
# Hidden layer 1: 64 nodes
# Optimizer: Adam
# Loss: Binary Crossentropy
# Metrics: Accuracy
# Epochs: 100
# Batch size: 10

In [ ]:
model = Sequential()
model.add(Dense(32, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1, activation = 'sigmoid'))
model.compile(optimizer='adam', loss = 'binary_crossentropy', metrics=['accuracy'])
model.fit(X_train, y_train.to_numpy(), epochs = 100, batch_size = 10, verbose=1) 

In [ ]:
y_pred = model.predict_classes(X_test)
model.evaluate(X_test, y_test.to_numpy())

In [ ]:
y_pred

In [ ]:
confusion_matrix(y_test, y_pred)

In [ ]:
accuracy_score(y_test, y_pred)

# Insights derived from altering model hyperparameters:

- A smaller batch size produced more accurate results, but took longer to run.

- Adding more nodes in the hidden layers did not seem to improve the accuracy of the model.

- Increasing the number of epochs decreased the accuracy.

- My best model was number one:
 
 Model#1
 
 Input layer: 32 nodes
 
 Hidden layer 1: 64 nodes
 
 Optimizer: Adam
 
 Loss: Binary Crossentropy
 
 Metrics: Accuracy
 
 Epochs: 10
 
 Batch size: 10
 
True negatives: 1514 (Those who my model predicted would not exit, who did not actually exit)

False negatives: 79 (Those who my model predicted would exit, but who did not actually exit)

False positive: 200 (Those who my model predicted would not exit, but who did actually exit)

True positives: 207 (Those who my model predicted would exit the bank, who did actually exit the bank)

 accuracy of model: 86%